### DISCUSSION TITLE
Serializable Async methods

### DISCUSSION THREAD

**Discussion by jmaine** [dotnet/csharplang#1582] [Language Ideas]
I would like serializable async methods built into C#.

https://blogs.msdn.microsoft.com/lucian/2012/11/23/how-to-hibernate-async-methods-how-to-serialize-task/

**Comment by HaloFour** [comment]
Could this be achieved today with AsyncMethodBuilder?

https://blogs.msdn.microsoft.com/seteplia/2018/01/11/extending-the-async-methods-in-c/

**Comment by jmaine** [comment]
Most of the infrastructure is there; however, the main issue is (de)serializing the state machine. All the fields/properties on the compiler generated state machine must be stable from save to resume.

**Comment by jnm2** [comment]
What would happen if the method had a non-serializable local or parameter?

**Comment by scalablecory** [comment]
Done right, this could bring us into Erlang territory of robust scale-out. @jnm2's point would severely limit when it could be used, though.

**Comment by jmaine** [comment]
There are other use cases for this, also.

**Comment by HaloFour** [comment]
I wonder how [Azure Durable Functions](https://docs.microsoft.com/en-us/azure/azure-functions/durable-functions-overview) handle this as they seem to support stateful long-lived asynchronous workflows using `async`/`await` in C#.

**Comment by svick** [comment]
@HaloFour They restart the whole function. From the article you linked to:

> Once an orchestration function is given more work to do […], the orchestrator wakes up again and re-executes the entire function from the start in order to rebuild the local state. If during this replay the code tries to call a function (or do any other async work), the Durable Task Framework consults with the *execution history* of the current orchestration. If it finds that the activity function has already executed and yielded some result, it replays that function's result, and the orchestrator code continues running.

This requires the code to be deterministic:

> The replay behavior creates constraints on the type of code that can be written in an orchestrator function. For example, orchestrator code must be deterministic, as it will be replayed multiple times and must produce the same result each time.

[And to notify the framework about any `async` work:](https://docs.microsoft.com/en-us/azure/azure-functions/durable-functions-checkpointing-and-replay#orchestrator-code-constraints)

> Orchestrator code must **never initiate any async operation** except by using the DurableOrchestrationContext API. For example, no `Task.Run`, `Task.Delay` or `HttpClient.SendAsync`.

**Comment by kind-serge** [comment]
**I have a more concrete proposal.**

As a starter, add to the framework:
1. Low level API to access state machine fields with their metadata.
1. Low level API to control execution of generated state machines.

Exact serialization mechanism should not be built-in into the language.

**Why?**

In short, it is an incredibly powerful solution for modern cloud distributed systems. You can find the complete explanation in [Conquest of Distributed Systems (part 4) — The Now and the Vision](https://medium.com/@sergiis/conquest-of-distributed-systems-part-4-the-now-and-the-vision-e844c9aee2c7).

**Proof**
The [D-ASYNC project](https://github.com/Dasync/Dasync) implements the vision, but has to rely on internal implementation details.

_P.S. I have to disclose that I'm the author of the content provided by given references._

**Comment by isral** [comment]
I want serializable IEnumerator for save and load game coroutine.
For who doesn't care about game version, just serialize it.
For who care, user must provide suspension point id on yield and disallow same name local variable.

**Comment by IbrahimElshafey** [comment]
I created a library that may help us with scenarios where we want to separate man functionality from the main workflow:
[Resumable Functions](https://github.com/IbrahimElshafey/ResumableFunctions) package will enable you to write code like the below:
```C#
[ResumableFunctionEntryPoint("RequestApprovalWorkflow.RequestApprovalFlow")]
internal async IAsyncEnumerable<Wait> RequestApprovalFlow()
{
    //wait for a method to be executed, If the match expression satisfied it will continue
    //method may be called in at any time
    yield return Wait<Request, bool>(_service.UserSubmitRequest, WaitSubmitRequest)
            .MatchIf((request, result) => request.Id > 0)
            .SetData((request, result) => UserRequest == request);

    //save state in the class contains the resumable function
    ManagerApprovalTaskId = _service.AskManagerApproval(UserRequest.Id);
    
    //wait for another new method
    yield return Wait<ApproveRequestArgs, int>(_service.ManagerApproval, "Wait Manager Approval")
            .MatchIf((approveRequestArgs, approvalId) => approvalId > 0 && approveRequestArgs.TaskId == ManagerApprovalTaskId)
            .SetData((approveRequestArgs, approvalId) => ManagerApprovalResult == approveRequestArgs);

   //some other code ....
}
```
